src/callbacks_step/amr.jl

```diff
@@ -282,14 +282,92 @@ function (amr_callback::AMRCallback)(u_ode::AbstractVector, mesh::TreeMesh,
     coarsened_original_cells = Int[]
   end
 
- # Return true if there were any cells coarsened or refined, otherwise false
- has_changed = !isempty(refined_original_cells) || !isempty(coarsened_original_cells)
- if has_changed # TODO: Taal decide, where shall we set this?
-  # don't set it to has_changed since there can be changes from earlier calls
-  mesh.unsaved_changes = true
- end
-
- return has_changed
+  # Return true if there were any cells coarsened or refined, otherwise false
+  has_changed = !isempty(refined_original_cells) || !isempty(coarsened_original_cells)
+  if has_changed # TODO: Taal decide, where shall we set this?
+    # don't set it to has_changed since there can be changes from earlier calls
+    mesh.unsaved_changes = true
+  end
+
+  if has_changed && mpi_isparallel()
+    old_mpi_ranks_per_cell = copy(mesh.tree.mpi_ranks)
+    # Load balancing
+    partition!(mesh)
+
+    # Retain current solution data
+    old_n_elements = nelements(dg, cache)
+    old_cell_ids = copy(cache.elements.cell_ids)
+    old_u_ode = copy(u_ode)
+    GC.@preserve old_u_ode begin # OBS! If we don't GC.@preserve old_u_ode, it might be GC'ed
+      old_u = wrap_array(old_u_ode, mesh, equations, dg, cache)
+
+      # Get new list of leaf cells
+      leaf_cell_ids = local_leaf_cells(mesh.tree)
+
+      # re-initialize elements container
+      @unpack elements = cache
+      resize!(elements, length(leaf_cell_ids))
+      init_elements!(elements, leaf_cell_ids, mesh, dg.basis.nodes)
+
+      resize!(u_ode, nvariables(equations) * nnodes(dg)^ndims(mesh) * nelements(dg, cache))
+      u = wrap_array(u_ode, mesh, equations, dg, cache)
+
+      # Find elements that will change their rank and send their data to the new rank
+      for old_element_id in 1:old_n_elements
+        cell_id = old_cell_ids[old_element_id]
+        if cell_id ∉ leaf_cell_ids
+          # Send element data to new rank, use cell_id as tag (non-blocking)
+          dest = mesh.tree.mpi_ranks[cell_id]
+          MPI.Isend(vec(old_u[:, .., old_element_id]), dest, cell_id, mpi_comm())
```

Comment:
When you use non-blocking communication, you always need to terminate the corresponding request. [`MPI.Isend`](https://juliaparallel.github.io/MPI.jl/latest/pointtopoint/#MPI.Isend) will return an [`MPI.Request`](https://juliaparallel.github.io/MPI.jl/latest/pointtopoint/#MPI.Request), which needs to be stored. Then, after all communication is finished, you need to [`MPI.Waitall!`](https://juliaparallel.github.io/MPI.jl/latest/pointtopoint/#MPI.Waitall!) on all requests until they are finished.

I think a good point for waiting on the requests would be after the next loop with the blocking `MPI.Recv`s - once they are processed, `MPI.Waitall!` is likely to return immediately, since most of the sending will have already been processed by then.